src/index.ts

```diff
@@ -0,0 +1,174 @@
+export interface VisualObserverEntry {
+  target: Element;
+  contentRect: DOMRectReadOnly;
+  isAppearing: boolean;
+}
+
+export interface VisualObserverCallback {
+  (this: VisualObserver, entries: VisualObserverEntry[], observer: VisualObserver): void;
+}
+
+interface VisualObserverElement {
+  io: IntersectionObserver | null;
+  threshold: number;
+  isFirstUpdate: boolean;
+}
+
+const floor = Math.floor;
+
+const root = document.documentElement;
+
+/**
+ * Create an observer that notifies when an element is resized, moved, or added/removed from the DOM.
+ */
+export class VisualObserver {
+  #callback: VisualObserverCallback;
```

Comment:
We can do this later, but my increasing view is that private fields are generally harmful (due to the `Proxy` issue). If we're in TS, you can just make them `private`.

Reply:
Fair points I can underscore and make them`private`/`protected`